### PR TITLE
Prevent rocksdb segfaults when accessing closed storage

### PR DESCRIPTION
**Pull request by garyschulte**
## PR description
gate transaction access on whether the underlying storage has closed.  This prevents segfaults at shutdown when we close rocksdb, but there are async processes attempting to access the db.

Currently evaluating to ensure there is no noticeable performance regression.

## Fixed Issue(s)

#5362

### CODE REVIEW

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java [hunk 1]
```diff
@@ -114,4 +114,11 @@ public interface KeyValueStorage extends Closeable {
    * @throws StorageException problem encountered when starting a new transaction.
    */
   KeyValueStorageTransaction startTransaction() throws StorageException;
+
+  /**
+   * Return Whether the underlying storage is closed or not.
```
Comment (macfarla) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java:119]:
```suggestion
   * Return Whether the underlying storage is closed.
```

File: services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java [hunk 2]
```diff
@@ -151,8 +163,20 @@ public void commit() throws StorageException {
...
+
   @Override
   public SnappedKeyValueStorage clone() {
     return new LayeredKeyValueStorage(hashValueStore, parent);
   }
+
+  private void throwIfClosed() {
+    if (parent.isClosed()) {
+      LOG.error("Attempting to use a closed RocksDBKeyValueStorage");
+      throw new IllegalStateException("Storage has been closed");
```
Comment (macfarla) [inline services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java:179]: RocksDBSnapshotTransaction uses StorageException - does it make sense to match that here?
      throw new StorageException("Storage has already been closed");

File: services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorage.java [hunk 3]
```diff
@@ -123,6 +123,13 @@ default boolean containsKey(final S segment, final byte[] key) throws StorageExc
    */
   void clear(S segmentHandle);
 
+  /**
+   * Whether the underlying storage is closed or not.
```
Comment (macfarla) [inline services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorage.java:127]:
```suggestion
   * Whether the underlying storage is closed.
```

File: services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorage.java [hunk 4]
```diff
@@ -123,6 +123,13 @@ default boolean containsKey(final S segment, final byte[] key) throws StorageExc
...
+   *
+   * @return boolean indicating whether the underlying storage is closed or not.
```
Comment (macfarla) [inline services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorage.java:129]:
```suggestion
   * @return boolean indicating whether the underlying storage is closed.
```

File: services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorageAdapter.java [hunk 5]
```diff
@@ -101,23 +112,39 @@ public KeyValueStorageTransaction startTransaction() throws StorageException {
...
+
+  @Override
+  public boolean isClosed() {
+    return storage.isClosed();
+  }
+
+  private void throwIfClosed() {
+    if (storage.isClosed()) {
+      LOG.error("Attempting to use a closed Storage instance.");
+      throw new IllegalStateException("Storage has been closed");
```
Comment (macfarla) [inline services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedKeyValueStorageAdapter.java:147]: IllegalStateEx vs StorageEx again
      throw new StorageException("Storage has already been closed");

**Review by macfarla — APPROVED** [review]
LGTM, couple of nits

File: services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java (hunk 2 above)
Comment (garyschulte) [reply services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java:179]: Yes, I think storage exception is better everywhere possible :+1:

**Pull request merged**